### PR TITLE
fix(aws-events): restrict eventbus statementId to 64 characters

### LINKED ISSUE #22120: (events): ValidationError when creating an EventBus that has CrossAccount access.
### Describe the bug

Basically, the id that's created for the support stack statementId is 5 characters too long (69 characters). It then fails validation and the stack cannot be deployed. 

It looks like this has been reported before: https://github.com/aws/aws-cdk/issues/19941.

### Expected Behavior

The stack deploys as expected.

### Current Behavior

Error Message: `Stack Deployments Failed: Error: The stack named ci-cd-EventBusPolicy-support-us-west-2-{account2} failed creation, it may need to be manually deleted from the AWS console: ROLLBACK_COMPLETE: 1 validation error detected: Value 'Allow-account-{account2}-c884c8876055cffba97afb1bc5a28125a7cac73762' at 'statementId' failed to satisfy constraint: Member must have length less than or equal to 64 (Service: AWSEvents; Status Code: 400; Error Code: ValidationException; Request ID: c521bf0c-abec-4042-90ed-54823e58a58e; Proxy: null)`

### Reproduction Steps

```import { App, aws_codecommit } from 'aws-cdk-lib';
import { Stack, StackProps } from 'aws-cdk-lib';
import { IRepository } from 'aws-cdk-lib/aws-codecommit';
import {
  CodePipeline,
  CodePipelineSource,
  ShellStep,
} from 'aws-cdk-lib/pipelines';
import { Construct } from 'constructs';

export interface ReproStackProps extends StackProps {
  codeRepository: IRepository;
}

export class CodeStack extends Stack {
  public codeRepository: aws_codecommit.IRepository;
  constructor(scope: Construct, props: StackProps) {
    super(scope, 'CodeStack', props);
    this.codeRepository = aws_codecommit.Repository.fromRepositoryName(
      this,
      'code-repo',
      'repo',
    );
  }
}

export class ReproStack extends Stack {
  constructor(scope: Construct, props: ReproStackProps) {
    super(scope, 'repro', props);

    new CodePipeline(this, 'code-pipeline', {
      crossAccountKeys: true,
      synth: new ShellStep('synth', {
        input: CodePipelineSource.codeCommit(props.codeRepository, 'mainline'),
        commands: ['yarn install', 'yarn build'],
      }),
    });
  }
}
const app = new App();

const codeStack = new CodeStack(app, {
  env: {
    region: 'us-west-2',
    account: 'account1',
  },
});

new ReproStack(app, {
  env: {
    region: 'us-west-2',
    account: 'account2',
  },
  codeRepository: codeStack.codeRepository,
});
```

### Possible Solution

When trying to upgrade past version 1.150 we ran into ValidationException while trying to create a EventBusPolicy cross account. This failure is probably happening because of the following line

statementId: `Allow-account-${sourceAccount}-${this.node.addr}`,

It can be found here: https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-events/lib/rule.ts#L400.

### Additional Information/Context

_No response_

### CDK CLI Version

2.41.0

### Framework Version

_No response_

### Node.js Version

v16.17.0

### OS

MacOS Monterey 12.4

### Language

Typescript

### Language Version

Version 4.8.3

### Other information

_No response_

### PR DESCRIPTION
**Pull request by sennyeya**
Fixes #22120, #21808.

Current setup does not allow deployment of the EventBus support stack due to StatementId being larger than 64 characters.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-events/test/rule.test.ts [hunk 1]
```diff
@@ -1015,7 +1015,7 @@ describe('rule', () => {
       const eventBusPolicyStack = app.node.findChild(`EventBusPolicy-${sourceAccount}-us-west-2-${targetAccount}`) as cdk.Stack;
       Template.fromStack(eventBusPolicyStack).hasResourceProperties('AWS::Events::EventBusPolicy', {
         'Action': 'events:PutEvents',
-        'StatementId': `Allow-account-${sourceAccount}-${nodeAddr}`,
+        'StatementId': `Allow-account-${sourceAccount}-${nodeAddr}`.substring(0, 64),
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-events/test/rule.test.ts:1018]: Please test against the raw value you expect here, not against a function.

File: packages/@aws-cdk/aws-events/lib/rule.ts [hunk 2]
```diff
@@ -370,7 +370,7 @@ export class Rule extends Resource implements IRule {
         });
         new CfnEventBusPolicy(eventBusPolicyStack, 'GivePermToOtherAccount', {
           action: 'events:PutEvents',
-          statementId: `Allow-account-${sourceAccount}-${this.node.addr}`,
+          statementId: `Allow-account-${sourceAccount}-${this.node.addr}`.substring(0, 64),
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-events/lib/rule.ts:373]: What happens if this cuts these off in a way that there will be a duplicate statementId? Will this cause collisions of any sort?

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-events/lib/rule.ts (hunk 2 above)
Comment (sennyeya) [reply packages/@aws-cdk/aws-events/lib/rule.ts:373]: That was my thinking as well, but discussed this on the linked ticket and there should be precedence for this solution.

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-events/lib/rule.ts:373]: The precedence that @peterwoodworth referred to uses `Names.uniqueResourceName()` to ensure there are no collisions. I don't think he meant that doing it this raw way was the way to go.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Please see inline comment. Also, per PR Linter, we need integ tests.

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by sennyeya** [comment]
> Please see inline comment. Also, per PR Linter, we need integ tests.

Just wanting to verify, what would the integ tests be testing? I have no experience in the EventBridge space.

**Comment by TheRealAmazonKendra** [comment]
> > Please see inline comment. Also, per PR Linter, we need integ tests.
> 
> Just wanting to verify, what would the integ tests be testing? I have no experience in the EventBridge space.

Basically, this never could have deployed. We need a test that  includes this functionality to prove that the deployment succeeds. This one is tricky because it requires cross account tests so rather than a fully functional test, we'd need steps taken to manually verify the test. As an example of this, see @corymhall's integ test here: https://github.com/aws/aws-cdk/pull/22122/files#diff-106ed5ff6c61e370348bf17a02cb29ccc3d438c6af02ef2f70453cd3e73a237e

**Comment by sennyeya** [comment]
@TheRealAmazonKendra Just following up on this, I managed to setup an integ test that does what I want it to do but it's failing at 
```
eventVerification.assertAtPath('Policy.Statement.0.Sid', ExpectedResult.stringLikeRegexp( 'Allow-account-'));
```
with the error message:
```
CustomResource attribute error: Vendor response doesn't contain apiCallResponse.Policy.Statement.0.Sid key in object arn:aws:cloudformation:us-east-1:ACCOUNT:stack/CrossAccountDeployDefaultTestDeployAssertB5328BEF/72124930-4f37-11ed-8394-0ec240785f9f|AwsApiCallEventBridgedescribeEventBus|1062e942-73a5-450d-93f4-ac0dddadd57d in S3 bucket cloudformation-custom-resource-storage-useast1
```

Even though the logs show:
```
"Status": "SUCCESS",
    "Reason": "OK",
    "PhysicalResourceId": "AwsApiCallEventBridgedescribeEventBus",
    "StackId": "arn:aws:cloudformation:us-east-1:ACCOUNT:stack/CrossAccountDeployDefaultTestDeployAssertB5328BEF/72124930-4f37-11ed-8394-0ec240785f9f",
    "RequestId": "7de3f5fe-fec8-4253-80b3-a6706fa0f3f3",
    "LogicalResourceId": "AwsApiCallEventBridgedescribeEventBus",
    "NoEcho": false,
    "Data": {
        "apiCallResponse.Name": "default",
        "apiCallResponse.Arn": "arn:aws:events:us-east-1:546328076441:event-bus/default",
        "apiCallResponse.Policy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"Allow-account-SECOND_ACCOUNT-FromCrossAccouuleStackMyRule68A189ED\",\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::SECOND_ACCOUNT:root\"},\"Action\":\"events:PutEvents\",\"Resource\":\"arn:aws:events:us-east-1:ACCOUNT:event-bus/default\"}]}"
    }
```

Any ideas on what's causing the error message? I think the JSON notation (Policy.Statement.0.Sid) is correct?

**Comment by TheRealAmazonKendra** [comment]
> @TheRealAmazonKendra Just following up on this, I managed to setup an integ test that does what I want it to do but it's failing at
> 
> ```
> eventVerification.assertAtPath('Policy.Statement.0.Sid', ExpectedResult.stringLikeRegexp( 'Allow-account-'));
> ```
> 
> with the error message:
> 
> ```
> CustomResource attribute error: Vendor response doesn't contain apiCallResponse.Policy.Statement.0.Sid key in object arn:aws:cloudformation:us-east-1:ACCOUNT:stack/CrossAccountDeployDefaultTestDeployAssertB5328BEF/72124930-4f37-11ed-8394-0ec240785f9f|AwsApiCallEventBridgedescribeEventBus|1062e942-73a5-450d-93f4-ac0dddadd57d in S3 bucket cloudformation-custom-resource-storage-useast1
> ```
> 
> Even though the logs show:
> 
> ```
> "Status": "SUCCESS",
>     "Reason": "OK",
>     "PhysicalResourceId": "AwsApiCallEventBridgedescribeEventBus",
>     "StackId": "arn:aws:cloudformation:us-east-1:ACCOUNT:stack/CrossAccountDeployDefaultTestDeployAssertB5328BEF/72124930-4f37-11ed-8394-0ec240785f9f",
>     "RequestId": "7de3f5fe-fec8-4253-80b3-a6706fa0f3f3",
>     "LogicalResourceId": "AwsApiCallEventBridgedescribeEventBus",
>     "NoEcho": false,
>     "Data": {
>         "apiCallResponse.Name": "default",
>         "apiCallResponse.Arn": "arn:aws:events:us-east-1:546328076441:event-bus/default",
>         "apiCallResponse.Policy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"Allow-account-SECOND_ACCOUNT-FromCrossAccouuleStackMyRule68A189ED\",\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::SECOND_ACCOUNT:root\"},\"Action\":\"events:PutEvents\",\"Resource\":\"arn:aws:events:us-east-1:ACCOUNT:event-bus/default\"}]}"
>     }
> ```
> 
> Any ideas on what's causing the error message? I think the JSON notation (Policy.Statement.0.Sid) is correct?

It's been flattened as a string. You'd want to look for 
```
eventVerification.assertAtPath('Policy', ExpectedResult.stringLikeRegexp( 'Allow-account-'));
```
I think you could also just do 
```
eventVerification.expect(ExpectedResult.objectLike({
  Policy: {
    Version: '2021-10-17',
    Statement: [
      {
        Sid: Match.stringLikeRegexp('Allow-acount-'),
        Effect: 'Allow',
        Principal: {
          AWS: Match.stringLikeRegexp('arn:aws;iam::')
        },
        Action: 'events:PutEvents',
        Resource: Match.stringLikeRegexp('arn:aws:events:us-east-1:')
      }
    ]
  }
}
```
This may not be precisely correct because I may have misplaces a comma or not gotten the json quite right, but this general format should get you there.

**Review by TheRealAmazonKendra — APPROVED** [review]

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Review by TheRealAmazonKendra — DISMISSED** [review]

**Review by TheRealAmazonKendra — APPROVED** [review]

**Pull request merged**